Add wasm32-unknown-emscripten platform support document

juntyr opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

This PR adds the platform support document for wasm32-unknown-emscripten, and adds a warning about breaks in Emscripten ABI compatibility (see #131467).

I mostly based the document off the wasm32-unknown-unknown docs and some of the information may still be missing (e.g. who's the target maintainer) or outdated (e.g. the build requirements). I still hope that it provides a good starting point.

r? @workingjubilee

juntyr committed:
Add wasm32-unknown-emscripten platform support document

Co-authored-by: Hood Chatham <roberthoodchatham@gmail.com>
Co-authored-by: Alex Crichton <alex@alexcrichton.com>
Co-authored-by: Kleis Auke Wolthuizen <github@kleisauke.nl>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11302957320/job/31439514460) [(plain)](https://github.com/rust-lang/rust/commit/262b381926ed98473943c26dbb6c0648ff7af2c2/checks/31439514460/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.901 Building wheels for collected packages: reuse
#13 2.902   Building wheel for reuse (pyproject.toml): started
#13 3.155   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.156   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.157   Stored in directory: /tmp/pip-ephem-wheel-cache-6bgfrg0k/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.159 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.575 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.575 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.121 Collecting virtualenv
#13 4.121 Collecting virtualenv
#13 4.179   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.468      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 20.9 MB/s eta 0:00:00
#13 4.531 Collecting filelock<4,>=3.12.2
#13 4.540   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.576 Collecting platformdirs<5,>=3.9.1
#13 4.586   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.607 Collecting distlib<1,>=0.3.7
#13 4.632      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 36.0 MB/s eta 0:00:00
#13 4.632      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 36.0 MB/s eta 0:00:00
#13 4.713 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.905 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 5.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      206784 kB
DirectMap2M:     8181760 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARNING: `rust.download-rustc` is enabled. The `rust.channel` option will be overridden by the CI rustc's channel.
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5598 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:25: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.2)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Oct 12 05:17:37 UTC 2024
  network time: Sat, 12 Oct 2024 05:17:37 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented:
> e.g. who's the target maintainer

I tried to figure this out before and it seems like we don't really have one. It doesn't seem to have been much of a problem for this target (presumably overlap with the other wasm targets helps) but we should definitely try to find somebody willing to put their name here. Of course this doesn't block improving the docs, this is needed regardless.

@hoodmane and @tlively you seem to have done a lot of the work for this target, any interest in being added as a target maintainer? The bar is very low, it mostly means you are willing to be pinged when anyone has emscripten-specific questions.

tgross35 commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,98 @@
+
+## Requirements
+
+This target is cross-compiled. The Emscripten compiler toolchain must be
+installed to link Rust-compiled code with Emscripten's libc and other C/C++
+code.
```
Something here should probably link to https://emscripten.org/, https://emscripten.org/docs/getting_started/downloads.html, or the relevant github repos.

juntyr replied:
Done

tgross35 commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,98 @@
+[`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target, which only has
+partial support for the Rust standard library, `wasm32-unknown-emscripten`
+integrates with JavaScipt on the web to provide a more featureful standard library.
+This target is often used when compiling for the web, accessing web APIs, or
+interoperating with C and C++ code.
```
Should we mention something about preferring `wasm32-unknown-unknown` or the `wasi` targets if C/C++ interop isn't needed? I'm not sure what kind of usecases would encourage using one over the other, but this would be great to document to steer people to the right thing.

Cc @alexcrichton

juntyr replied:
I've added two paragraphs at the end of the section but would also appreciate Alex's input.

alexcrichton replied:
I might recommend rephrasing/reorganizing this section along the lines of:

1. What is Emscripten and how this target integrates? (e.g. Emscripten is a C/C++ toolchain not related to Rust used for running C/C++ code on the web)
2. Why Emscripten is used and how Rust integrates. (e.g. Emscripten implements POSIX in terms of web APIs and provides a C standard library for that)
3. How this contrasts to other wasm targets in Rust (e.g. unlike `wasm32-unknown-unknown` it can run C/C++ and provides a full implementation of libc/libstd, unlike `wasm32-wasip1` it requires JS glue code to run on the web because the interface between wasm/JS isn't standard)

In terms of C/C++ I'd recommend focusing on both that and library support for Emscripten. Emscripten isn't just about C/C++ but also all the polyfills provided for libraries like SDL (IIRC). The contrast to `wasm32-unknown-unknown` can indicate that the unknown target has no support for C/C++ or anything that requires JS (or at least not inherently).

When comparing to `wasm32-wasip1` I'd recommend mostly contrasting it in terms of web support. The WASI target has no standard way of accessing web APIs where Emscripten has the ability to run arbitrary JS from wasm and access many web APIs. In the end the difference here will likely boil down to "are you running on the web or not" but that doesn't alone accurately capture the nuance of you can run WASI on the web but you just won't have access to the full suite of web APIs by default.

---

That may be more words than desired, sorry about that! Overall I think the wording is fine as-is, so what I'm thinking here is just minor updates here and there. Feel free to take or leave any suggestions as you see fit!

juntyr replied:
Thank you for your extensive feedback @alexcrichton, I really appreciate it!

I've tried to address it in the latest commit, I hope I managed to capture your feedback

tgross35 commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,98 @@
+This target is often used when compiling for the web, accessing web APIs, or
+interoperating with C and C++ code.
+
+One existing user of this target (please feel free to edit and expand this list
+too) is the [`pyodide` project](https://github.com/pyodide/pyodide) which provides
```
Not sure if the `please feel free to edit and expand this list too` part was intentional but it can probably be removed.

juntyr replied:
Done

hoodmane commented:
I would like to be pinged with emscripten related questions. I would also suggest @sbc100 should be kept in the loop.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11311910803/job/31458690107) [(plain)](https://github.com/rust-lang/rust/commit/cfb6ccc53163368b7eb3fb4a450365a371d139ca/checks/31458690107/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.79s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.13s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustc/print.html:9926: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/print.html:15968: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support.html:347: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support/wasm32-unknown-unknown.html:221: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
number of HTML files scanned: 43689
number of HTML redirects found: 13733
number of links checked: 3206901
number of links ignored due to external: 82379
number of links ignored due to external: 82379
number of links ignored due to exceptions: 17
number of intra doc links ignored: 8
errors found: 4
found some broken links
Command has failed. Rerun with -v to see more details.
  local time: Sun Oct 13 06:53:38 UTC 2024
  network time: Sun, 13 Oct 2024 06:53:38 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11313094172/job/31461437776) [(plain)](https://github.com/rust-lang/rust/commit/a1ed63b06e3f82669ea89c5fb8958c81661c6027/checks/31461437776/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.83s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.13s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustc/print.html:9926: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/print.html:15968: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support.html:347: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support/wasm32-unknown-unknown.html:221: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
number of HTML files scanned: 43688
number of HTML redirects found: 13733
number of links checked: 3205506
number of links ignored due to external: 82355
number of links ignored due to external: 82355
number of links ignored due to exceptions: 17
number of intra doc links ignored: 8
errors found: 4
found some broken links
Command has failed. Rerun with -v to see more details.
  local time: Sun Oct 13 10:02:03 UTC 2024
  network time: Sun, 13 Oct 2024 10:02:03 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

juntyr commented:
I'm unfortunately not sure which links are broken, at least on GitHub all new links on my branch work

alexcrichton commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
````diff
@@ -0,0 +1,109 @@
+#[cfg(all(target_family = "wasm", target_os = "emscripten"))]
+```
+
+Note that there is no way to tell via `#[cfg]` whether code will be running on
+the web or not.
````
This clause may not be so applicable for the emscripten target since by definition almost all emscripten code runs on the web

juntyr replied:
Removed - done

alexcrichton commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,109 @@
+
+## Enabled WebAssembly features
+
+WebAssembly is an evolving standard which adds new features such as new
+instructions over time. The `wasm32-unknown-emscripten` target inherits the
+default settings of Emscripten.
```
This might want to be clarified a bit because it's not quite true. The default features are inherited from LLVM that rustc uses which may differ from Emscripten depending on the user's installation of the Emscripten toolchain. In that sense I'd recommend pointing to the `wasm32-unknown-unknown` docs about features here because the documentation there will apply to the Emscripten target here too.

juntyr replied:
I added a reference to the `wasm32-unknown-unknown` docs

alexcrichton commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,109 @@
+
+## Target maintainers
+
+When this target was added to the compiler, platform-specific documentation here
+was not maintained at that time. This means that the list below is not
+exhaustive, and there are more interested parties in this target. That being
+said, those interested in maintaining this target are:
```
I realize this is probably copy/pasted from the `wasm32-unknown-unknown` target but given the set of three maintainers below you can probably drop this clause. I mostly added it to the unknown target because I was the only one listed and wanted to clarify why there was just one.

juntyr replied:
Done

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11316509367/job/31468867405) [(plain)](https://github.com/rust-lang/rust/commit/d7f65a2cebb71a7ae55f506d96db26e7eac7564f/checks/31468867405/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.73s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.13s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustc/print.html:9926: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/print.html:15968: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support.html:347: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support/wasm32-unknown-unknown.html:221: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
number of HTML files scanned: 43688
number of HTML redirects found: 13733
number of links checked: 3205575
number of links ignored due to external: 82355
number of links ignored due to external: 82355
number of links ignored due to exceptions: 17
number of intra doc links ignored: 8
errors found: 4
found some broken links
Command has failed. Rerun with -v to see more details.
  local time: Sun Oct 13 18:46:20 UTC 2024
  network time: Sun, 13 Oct 2024 18:46:20 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
````diff
@@ -0,0 +1,116 @@
+$ cargo +nightly -Zbuild-std build
+```
+
+## Testing
+
+This target is not extensively tested in CI for the rust-lang/rust repository.
+It's recommended to test the [`wasm32-wasip1`](./wasm32-wasip1.md) target instead
+for WebAssembly compatibility.
````
Could you elaborate how contributors can test this target locally?

i.e. Tier 2 target policy says

> The target must provide documentation for the Rust community explaining [..] how to run tests for the target. If at all possible, this documentation should show how to run Rust programs and tests for the target using emulation, to allow anyone to do so. If the target cannot be feasibly emulated, the documentation should explain how to obtain and work with physical hardware, cloud systems, or equivalent.

see #131666, where Jubilee found that

> a fifth of the rustc UI test suite to fail on wasm32-unknown-emscripten when trying to run ./x.py test --target wasm32-unknown-emscripten

jieyouxu replied:
I was able to get past the test helpers stage with <https://emscripten.org/docs/getting_started/downloads.html> including to clone that repo, use `./emsdk` to fetch and activate the latest emcc toolchain, then use a provided shell script to setup the relevant env vars and make them available to bootstrap.

stage1 std seems to be wonky in the bootstrap test setup

```
warning: dropping unsupported crate type `dylib` for target `wasm32-unknown-emscripten`

warning: `std` (lib) generated 1 warning
```

ui tests failed due to linker-flavors and such and I didn't bother investigating.

juntyr replied:
@hoodmane Since you have gotten most tests to run, could you suggest some text?

hoodmane replied:
I've been running tests with:
```sh
./x.py test --target wasm32-unknown-emscripten --skip src/tools/linkchecker
```

hoodmane replied:
But you need `emcc` and `node` on the path first. To get `emcc` installed and on the path:
```sh
git clone https://github.com/emscripten-core/emsdk.git --depth 1
./emsdk/emsdk install 3.1.68
./emsdk/emsdk activate 3.1.68
source ./emsdk/emsdk_env.sh
```
And you can use whatever normal way to install node, but if we're going to recommend something, nvm is quite nice:
https://github.com/nvm-sh/nvm#install--update-script

juntyr replied:
Thanks! I've added those instructions

jieyouxu commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,116 @@
+
+Building this target can be done by:
+
+* Configure the `wasm32-unknown-emscripten` target to get built.
+* Configure LLD to be built.
+* Ensure the `WebAssembly` target backend is not disabled in LLVM.
+
+These are all controlled through `config.toml` options. It should be possible
+to build this target on any platform.
```
Suggestion: could you show a minimal `config.toml` snippet for this?

juntyr replied:
I've tried my best to add a minimal example config.toml

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
````diff
@@ -0,0 +1,116 @@
+It's recommended to conditionally compile code for this target with:
+
+```text
+#[cfg(all(target_family = "wasm", target_os = "emscripten"))]
````
I'd say it's sufficient to do:
```
#[cfg(target_os = "emscripten")]
```
It's often necessary to filter for wasm, not emscripten somewhere else:
```
#[cfg(all(target_family = "wasm", not(target_os = "emscripten)))]` 
```
I've hit a bunch of duplicate definition compile errors, though if people are building for Emscripten at all they would catch this one.

In a lot of cases the Emscripten branch is the same as the linux branch since Emscripten implements a lot of linux headers.

juntyr replied:
I've updated the text - thanks!

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,116 @@
+
+WebAssembly is an evolving standard which adds new features such as new
+instructions over time. This target's default set of supported WebAssembly
+features will additionally change over time. The `wasm32-unknown-emscripten` target
+inherits the default settings of LLVM which typically, but not necessarily, matches
+the default settings of Emscripten as well.
```
> inherits the default settings of LLVM

I would not say this, at least not at link time. `emcc` invokes the linker with a very large number of special options.

juntyr replied:
I've adapted the text

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,116 @@
+inherits the default settings of LLVM which typically, but not necessarily, matches
+the default settings of Emscripten as well.
+
+Please refer to the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md)
+target's documentation on which WebAssembly features Rust enables by default, how
+features can be disabled, and how Rust code can be conditionally compiled based on
+which features are enabled.
```
Over there, they list:
* `multivalue`
* `mutable-globals`
* `reference-types`
* `sign-ext`

These are all determined at compile time, so the advice over there should work. It does not apply as well to js/wasm interop features which are generally determined at link-time.

We might also note here that there is no support for `-fwasm-exceptions`, only for `-fexceptions` which will be added by default unless you compile with `panic=abort`. This is an important limitation. I would really like support for `wasm-exceptions`. 
https://github.com/rust-lang/rust/issues/112195

It would be nice to add wasm/js bigint interop (`-sWASM_BIGINT`) as another default feature. It's been around much longer than reference types and it makes a big improvement in code generation.

juntyr replied:
I agree with enabling the `-sWASM_BIGINT` feature. This should be tracked in a separate PR.

juntyr replied:
I've added a paragraph on the exception handling features.

hoodmane replied:
Opened a PR for `-sWASM_BIGINT`
https://github.com/rust-lang/rust/pull/131736

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,143 @@
+
+The `wasm32-unknown-emscripten` target is a WebAssembly compilation target which
+uses the [Emscripten](https://emscripten.org/) compiler toolchain. Emscripten is
+a C/C++ toolchain used for running C/C++ code on the web. It provides POSIX `libc`
+and `libstd` implementations, access to the OpenGL and SDL APIs, and the ability
```
It implements most of posix in a mostly compatible way, and also a lot of linux APIs. It might be worth saying here that the design goal is to allow C/C++ code (e.g., 3d games) written to target linux to be ported with as few changes as possible. Also, it uses musl libc.

juntyr replied:
I've adapted the wording.

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,157 @@
+
+## Requirements
+
+This target is cross-compiled. The Emscripten compiler toolchain (emsdk) must be
```
The emscripten compiler toolchain is `emcc`, `emsdk` is the "Emscripten software developer kit" which is primarily an installer and version manager for the emscripten compiler toolchain.

juntyr replied:
Thanks for catching that

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,157 @@
+<https://emscripten.org/docs/getting_started/downloads.html> on how to get started
+with the emsdk.
+
+You can also build the emsdk from source with:
```
This does not build from source. emsdk is just an installer for the toolchain.

Interestingly, [the code for the emscripten compiler toolchain](https://github.com/emscripten-core/emscripten/) itself has no compilation step, it is implemented in a mixture of Python and JavaScript. So you can just modify the source files in place and immediately have a patched compiler. The standard library / sysroot is compiled as needed (since `emcc` implements many many different ABIs it doesn't make sense to ship prebuilt static libraries, it builds them as needed and caches them).

All of the compiled code needed for implementing the compiler (as opposed to host wasm code) is upstream either in [binaryen](https://github.com/WebAssembly/binaryen) or in llvm itself.

juntyr replied:
Thanks for correcting my mistake - I've reworded the paragraph

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,157 @@
+WebAssembly using Emscripten and compiles Python extension modules written in Rust
+to the `wasm32-unknown-emscripten` target.
+
+If you want to generate a standalone bare-bones WebAssembly binary that does not
+require JS glue code and access to the web APIs and can thus run outside the web,
+the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target may be better
+suited for you. However, [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md)
+only has partial support for the Rust standard library and does not (easily)
+support interop with C/C++ code.
```
```suggestion
If you want to generate a standalone WebAssembly binary that does not
require access to the web APIs or the Rust standard library,
the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target may be better
suited for you. However, [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) does not (easily)
support interop with C/C++ code.
```
As an aside, someday someone should really add an option to make `wasm32-unknown-unknown` emit the clang ABI instead of the wasm-bindgen ABI to fix the interop problem...

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,157 @@
+
+**Tier: 2**
+
+The `wasm32-unknown-emscripten` target is a WebAssembly compilation target which
+uses the [Emscripten](https://emscripten.org/) compiler toolchain. Emscripten is
+a C/C++ toolchain designed to make it as easy as possible to port C/C++ code
+written for Linux to run on the web. It thus provides POSIX-compatible (musl) `libc`
+and `libstd` implementations and many Linux APIs, access to the OpenGL and SDL APIs,
+and the ability to run arbitrary JavaScript code, all based on web APIs using JS
+glue code. With the `wasm32-unknown-emscripten` target, Rust code can interoperate
+with Emscripten's ecosystem, C/C++ and JS code, and web APIs.
```
It would be nice if we could mention somewhere in here that emscripten code also can run in node or other JavaScript runtimes. (The goal of Emscripten is to run on web, but once you run on web it's not too hard to also run in other JS runtimes.)

juntyr replied:
Done

hoodmane reviewed (approved):
Thanks so much for writing this @juntyr. It looks great!

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11343280808/job/31545545512) [(plain)](https://github.com/rust-lang/rust/commit/eb47be3b3a63eb7e8cb9a2eb3ea9264fc510a2b6/checks/31545545512/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.80s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.13s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustc/print.html:9926: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/print.html:15968: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support.html:347: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support/wasm32-unknown-unknown.html:221: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
number of HTML files scanned: 43887
number of HTML redirects found: 13734
number of links checked: 3244029
number of links ignored due to external: 83085
number of links ignored due to external: 83085
number of links ignored due to exceptions: 17
number of intra doc links ignored: 8
errors found: 4
found some broken links
Command has failed. Rerun with -v to see more details.
  local time: Tue Oct 15 10:23:52 UTC 2024
  network time: Tue, 15 Oct 2024 10:23:52 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

juntyr commented:
@jieyouxu I've tried to address all feedback, could you take another look?

jieyouxu reviewed (approved):
Thanks, this LGTM now

jieyouxu commented:
> I would like to be pinged with emscripten related questions. I would also suggest @sbc100 should be kept in the loop.

If you want, you can setup a wasm/emscripten [notification group](https://rustc-dev-guide.rust-lang.org/notification-groups/about.html) and ask interested parties to join, so people can ping the group via `@rustbot ping emscripten` or `wasm`. Target-specific expertise is very welcomed if target experts can occasionally help to answer questions and such.

jieyouxu commented:
Not sure why linkchecking is failing though

juntyr commented:
> Not sure why linkcheck is failing though

I have no idea either. Is there any way to get the HTML files from CI to see which links are broken or do I need to investigate in a local build?

juntyr commented:
> > I would like to be pinged with emscripten related questions. I would also suggest @sbc100 should be kept in the loop.
> 
> If you want, you can setup a wasm/emscripten [notification group](https://rustc-dev-guide.rust-lang.org/notification-groups/about.html) and ask interested parties to join, so people can ping the group via `@rustbot ping emscripten` or `wasm`. Target-specific expertise is very welcomed if target experts can occasionally help to answer questions and such.

I’d also be interested in being pinged as well (for both wasm in general and emscripten). Should this be two groups or one with an alias (if that’s possible)? The docs show how to join one, how can a new group be created?

jieyouxu commented:
> I’d also be interested in being pinged as well (for both wasm in general and emscripten). Should this be two groups or one with an alias (if that’s possible)? The docs show how to join one, how can a new group be created?

The forge docs has steps for this: <https://forge.rust-lang.org/compiler/notification-groups.html>. It presently says it will need an MCP to create the ping groups (... and if you file one I'll just immediately second it anyway), but I'll ask the T-compiler leads if this is still current policy (i.e. if the MCP is needed).

jieyouxu commented:
Just to be explicit, @sbc100, are you willing to be added as a target maintainer for the `wasm32-unknown-emscripten` target?

jieyouxu commented:
@juntyr I think having a general-purpose wasm group and also an emscripten-specific ping group separately is a good idea. You'll need to create two marker teams over at the team repo. See https://forge.rust-lang.org/compiler/notification-groups.html#notification-groups for the steps.

The forge docs says a MCP is required, but I discussed the ping group matter with @davidtwco, one of our T-compiler co-leads, and we determined that a MCP is not required for this situation, so you can go ahead with the team PRs without needing a MCP.

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
```diff
@@ -0,0 +1,158 @@
+## Emscripten ABI Compatibility
+
+The Emscripten compiler toolchain does not follow a semantic versioning scheme
+that clearly indicates when breaking changes to the ABI can be made. If you are
+using the pre-compiled `std` from rustup, it may be compiled with a different
+Emscripten version than the local Emscripten you have installed. If a breaking
+change to the ABI was made in between the two versions, your code will exhibit
+undefined behaviour. If you come across issues, you can rebuild the Rust standard
+library with your local Emscripten version using
+
```
The situation is actually worse because Emscripten offers many different ABIs even for a single version of Emscripten depending on linker flags. To be ABI compatible, all of the linker flags that adjust ABIs must match. This includes things like `-fexceptions` and `-sWASM_BIGINT`. I think it would be good to be a bit more emphatic here. Best practices is either:
1. use `-Zbuild-std`, or
2. match the emscripten version bundled with Rust and be careful about any `-C link-arg`'s that you pass.

The third yolo option is not ideal...

juntyr replied:
I made some further adjustments to the text to highlight that almost everybody should use `-Zbuild-std`

jieyouxu commented:
I'm looking at the linkcheck failures to see if I can repro them locally.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11345014574/job/31550984467) [(plain)](https://github.com/rust-lang/rust/commit/02c052664f7d42db44d9d5fa45b57338764ceed8/checks/31550984467/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.76s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.13s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustc/print.html:9926: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/print.html:15968: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support.html:347: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
rustc/platform-support/wasm32-unknown-unknown.html:221: broken link - `rustc/platform-support/wasm32-unknown-emscripten.html`
number of HTML files scanned: 43887
number of HTML redirects found: 13734
number of links checked: 3244029
number of links ignored due to external: 83085
number of links ignored due to external: 83085
number of links ignored due to exceptions: 17
number of intra doc links ignored: 8
errors found: 4
found some broken links
Command has failed. Rerun with -v to see more details.
  local time: Tue Oct 15 12:14:17 UTC 2024
  network time: Tue, 15 Oct 2024 12:14:17 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
@juntyr linkchecker is failing because we need to declare the page exists in `SUMMARY.md`, lol.

juntyr commented:
> @juntyr linkchecker is failing because we need to declare the page exists in `SUMMARY.md`, lol.

Thanks @jieyouxu, that fixed it!

jieyouxu commented:
@juntyr thanks, this looks mostly ready to merge. Can you please squash the commits into one? Also, do you have contact with @sbc100, because I would like explicit confirmation from @sbc100 that they are happy to be listed as a target maintainer.

juntyr commented:
> @juntyr thanks, this looks mostly ready to merge. Can you please squash the commits into one? Also, do you have contact with @sbc100, because I would like explicit confirmation from @sbc100 that they are happy to be listed as a target maintainer. If you don't want to wait for their confirmation, then you can also drop them from maintainer list for now, and ask them to open a PR to add themselves in a follow-up PR.

@jieyouxu I've squashed the commits.

I had hoped we would get an answer from @sbc100, but since that's not the case I've removed @sbc100 from the maintainer list for the `wasm-unknown-emscripten` Rust target for now, which can always be easily changed in a future PR.

jieyouxu commented:
Thanks.
@bors r+ rollup

bors commented:
:pushpin: Commit 0804d1bc27bd9b64c1302851f047316f086a0d1a has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 0804d1bc27bd9b64c1302851f047316f086a0d1a -->
<!-- homu: {"type":"Approved","sha":"0804d1bc27bd9b64c1302851f047316f086a0d1a","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

hoodmane commented on src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md:
````diff
@@ -0,0 +1,168 @@
+library with your local Emscripten version and settings using:
+
+```sh
+cargo +nightly -Zbuild-std build
````
We should perhaps add a note here that this won't work for crates that use:
1. cargo vendor
2. panic=abort

You can work around the cargo vendor problem by building a custom sysroot with https://github.com/ralfJung/rustc-build-sysroot/ you still won't be able to build crates with `panic=abort` .
